### PR TITLE
Change of DigiToRaw code to keep SiStrip Error information during data repacking

### PR DESCRIPTION
**Pull request by rkunnawa**
This PR fixes the bug which caused empty histograms for SiStrip FED Errors in the DQM. This is prepared for the HIN data taking period after we saw the aforementioned issues. 
thanks! 
@abaty @icali

### CODE REVIEW

**Comment by abaty** [comment]
Some elaboration about this PR.  This code adds to the DigiToRaw repacking code the ability to copy the header of another rawDataCollector.  This is important for Heavy Ion data taking, as nearly all our data stored on disk is run through the repacker.  Without this fix we lose the information stored in the header about the error status of the SiStrips, and also are unable to fill some of the SiStrip FED DQM plots.

File: EventFilter/SiStripRawToDigi/plugins/SiStripDigiToRaw.cc [hunk 1]
```diff
@@ -74,36 +110,187 @@ namespace sistrip {
...
+	  }
+	  
+	  //need to construct full object to copy full header
+	  std::auto_ptr<const sistrip::FEDBuffer> fedbuffer;
+	  //std::auto_ptr<const sistrip::FEDBufferBase> bufferBase;
+	  try {
+	    fedbuffer.reset(new sistrip::FEDBuffer(rawfedData.data(),rawfedData.size(),true));
+	    //bufferBase.reset(new sistrip::FEDBufferBase(rawfedData.data(),rawfedData.size()));
+	  } catch (const cms::Exception& e) {
+	    std::cerr << "[sistrip::DigiToRaw::createFedBuffers_]"
```
Comment (slava77) [inline EventFilter/SiStripRawToDigi/plugins/SiStripDigiToRaw.cc]: use LogError or LogWarning

File: EventFilter/SiStripRawToDigi/plugins/SiStripDigiToRaw.cc [hunk 2]
```diff
@@ -74,36 +110,187 @@ namespace sistrip {
...
-
+	  
 	  // Determine APV pair number (needed only when using DetId)
 	  uint16_t ipair = ( useFedKey_ || mode_ == READOUT_MODE_SCOPE ) ? 0 : iconn->apvPairNumber();
-          
-          FEDStripData::ChannelData& chanData = fedData[iconn->fedCh()];
+	  
+	  FEDStripData::ChannelData& chanData = fedData[iconn->fedCh()];
 
+	  /*if ( edm::isDebugEnabled() ) {
```
Comment (slava77) [inline EventFilter/SiStripRawToDigi/plugins/SiStripDigiToRaw.cc]: please cleanup

File: EventFilter/SiStripRawToDigi/plugins/SiStripDigiToRawModule.cc [hunk 3]
```diff
@@ -21,10 +23,12 @@ namespace sistrip {
   DigiToRawModule::DigiToRawModule( const edm::ParameterSet& pset ) :
     inputModuleLabel_( pset.getParameter<std::string>( "InputModuleLabel" ) ),
     inputDigiLabel_( pset.getParameter<std::string>( "InputDigiLabel" ) ),
+    copyBufferHeader_(pset.getUntrackedParameter<bool>("CopyBufferHeader",false)),
     mode_( fedReadoutModeFromString(pset.getParameter<std::string>( "FedReadoutMode" ))),
     rawdigi_( false ),
     digiToRaw_(0),
-    eventCounter_(0)
+    eventCounter_(0),
+    rawDataTag_(pset.getUntrackedParameter<edm::InputTag>("RawDataTag",edm::InputTag("rawDataCollector","")))
```
Comment (slava77) [inline EventFilter/SiStripRawToDigi/plugins/SiStripDigiToRawModule.cc]: these should be tracked.

**Comment by slava77** [comment]
@rkunnawa 
please provide details how to test with a case where two fedRaw are read

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by rkunnawa** [comment]
@slava77 we do have the logWarning in the code. can you briefly explain what you wanted done? should we remove the  
+	  if ( edm::isDebugEnabled() ) {
thanks!

File: EventFilter/SiStripRawToDigi/plugins/SiStripDigiToRaw.cc (hunk 1 above)
Comment (slava77) [inline EventFilter/SiStripRawToDigi/plugins/SiStripDigiToRaw.cc]: line 143 here has "std::cerr" 
please change to LogWarning or LogError

**Comment by rkunnawa** [comment]
@slava77 thanks for pointing that out. im afraid i need to ask you again what you mean by " please provide details how to test with a case where two fedRaw are read". Do you want a config which we tested with the cout statements for a single fed (which you can find it here ~rkunnawa/public.testSiStripFEDMonitor_cfg.py on lxplus)? or im not sure?

**Comment by slava77** [comment]
@rkunnawa 
is this config what's supposed to run online or really just a small test setup?
I'd rather check the full setup.

**Comment by slava77** [comment]
@cmsbuild please test
[to have a test after the updates]

**Comment by rkunnawa** [comment]
@slava77 The config i sent you in the past is just a small test setup to print the headers. 
 Are we missing something else? @abaty is running HLT on some raw data files to check in parallel.

File: EventFilter/SiStripRawToDigi/plugins/SiStripDigiToRawModule.cc (hunk 3 above)
Comment (slava77) [inline EventFilter/SiStripRawToDigi/plugins/SiStripDigiToRawModule.cc]: Please comment or follow up: I think that RawDataTag and CopyBufferHeader should be tracked parameters

**Comment by slava77** [comment]
@rkunnawa 
please explain more explicitly what has been done here.
It looks like the digi2raw running in HLT (not anywhere else) in HI data taking has to add some more information to the repacked raw. 
And then this repacked raw will have some extra data useful only for DQM.

I don't see any differences in the tests done by cms-bot/jenkins. This probably means that the needed setup is not tested.

I wanted to get a configuration where the effects of this PR are actually visible.

**Comment by Martin-Grunewald** [comment]
Instead of messing about with untracked parameters, please code a proper fillDescriptions method,
esp. for the modules needed at HLT.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please test
maybe the DAS error goes away

**Comment by abaty** [comment]
I've attached the commands for running the workflow I am using to test this PR.  It consists of emulating the current HLT setup, and then running RECO and harvesting the DQM.  I have confirmed that the code copies the FED header by dumping and comparing both the buffers.  Unfortunately I have only been able to test on a few events, so I do not find any APV errors.   However, if I manually inject some errors into the buffer header they show up in the DQM (so this is not an issue with something in the RECO/DQM harvesting).  

Be advised that this test must run on virgin raw data that hasn't been repacked before (so only Heavy Ion data from 2010 or the 2015 HIPhysicsVirginRaw stream).

@slava77 

[PR_Test_Workflow.txt](https://github.com/cms-sw/cmssw/files/53338/PR_Test_Workflow.txt)

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@rkunnawa thank you for the instructions in https://github.com/cms-sw/cmssw/files/53338/PR_Test_Workflow.txt
 if I understand correctly, 
- this code is needed only online where the virginraw -> raw repacking is done
- new parameters should be known to the HLT configuration from condDb. The defaults should be provided via the fillDescriptions method mentioned by Martin already.  Without these changes, I don't think your updates can be propagated online where they are needed.

**Comment by rkunnawa** [comment]
Hi @slava77 (@abaty provided the instructions :) ) We are using this when the repacking is done at the HLT level and its specially for Heavy Ion data taking periods. 
@Martin-Grunewald  Also we are not familar with this fillDescriptions method that you are referring to. Can you please point us to an example so we can understand what is needed and try to code it here.

**Comment by icali** [comment]
@slava77 @Martin-Grunewald I believe that moving to the fillDescription method could be a solution and if not complicated it should be added. However, I would like to clarify that this is not a specific HLT module. We are talking of the standard tracker packer run all the time in any CMS data packing. The addition is to pack correctly the FEDheader (method never implemented correctly up to now). We are also using it in HLT and it is becoming a quite urgent fix. Without having the correct FED header propagated to the RAW data (now it is dropped in HLT), we can incur in tracking efficiency loss for miscalculation of "bad" modules/strips

**Comment by Martin-Grunewald** [comment]
@rkunnawa 
Just LXR search for ```fillDescription```.
One example is https://github.com/cms-sw/cmssw/blob/CMSSW_8_0_X/EventFilter/GctRawToDigi/plugins/GctRawToDigi.cc#L122

**Comment by Martin-Grunewald** [comment]
@icali 
If it is used in HLT HIon, it must have it.

**Comment by Martin-Grunewald** [comment]
See here for a description on a wiki:
https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp

**Comment by slava77** [comment]
On 12/7/15 8:05 AM, Ivan Amos Cali wrote:
> @slava77 <https://github.com/slava77> @Martin-Grunewald
> <https://github.com/Martin-Grunewald> I believe that moving to the
> fillDescription method could be a solution and if not complicated it
> should be added. However, I would like to clarify that this is not a
> specific HLT module. We are talking of the standard tracker packer run
> all the time in any CMS data packing. The addition is to pack correctly
> the FEDheader (method never implemented correctly up to now). We are
> also using it in HLT and it is becoming a quite urgent fix. Without
> having the correct FED header propagated to the RAW data (now it is
> dropped in HLT), we can incur in tracking efficiency loss for
> miscalculation of "bad" modules/strips
> 


Hi Ivan,

the default configuration for offline running of this module is not
changing any outputs
(at least based on what we monitor; if something is actually changing in
the default setup,
I'd like to get a pointer to that).
That's why I'm saying that the change is specific for online.

For the online configuration, code from this PR will just not run with
the current HLT config
(new parameters are not known). fillDescriptions should be implemented.
After that, I suppose, you will make changes in the HLT paths to set the
header copying.

**Comment by slava77** [comment]
@rkunnawa @abaty 
please post a message  when you are done updating this PR.

**Comment by abaty** [comment]
@slava77 
The fill description function you requested has been added; we are finished updating.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by Martin-Grunewald** [comment]
Are the parameter values used in fillDesciptions such that they do what is needed to fix the problem,
or are they set to to reproduce the old behaviour?

**Comment by slava77** [comment]
https://github.com/rkunnawa/cmssw/blob/SiStrigFED_DQMFix/EventFilter/SiStripRawToDigi/plugins/SiStripDigiToRawModule.cc#L28
these defaults should give the old behavior (CopyBufferHeader = false)

**Comment by Martin-Grunewald** [comment]
BTW, we'd also need a PR for 76X so that this feature does not get lost!
(80X presumably done thru automatic forward port of the 76X PR).

**Comment by slava77** [comment]
We have converged on this PR version close enough to request a 80X/76X PR.
Better have both (schedule of merging new code in 76X is not that clear)

**Comment by slava77** [comment]
actually, looking at 2010 HI workflow (140.51), there are differences : 
wf140 51_strip_errfrac
wf140 51_anyapv

left in 2D or black in 1D is the baseline (757patch3).
Do these plots mean that the errors were fake before or do we actually lose information now?

**Comment by amagnan** [comment]
Hi, the errors from before look the reasonable amount we expect, they are not fake, we should see the same ones with the new workflow. Were the options properly set (copyBufferHeader set to true and input rawdata collection name given as input to the digi2raw) ?

**Comment by slava77** [comment]
On 12/8/15 3:02 AM, amagnan wrote:
> Hi, the errors from before look the reasonable amount we expect, they
> are not fake, we should see the same ones with the new workflow. Were
> the options properly set (copyBufferHeader set to true and input rawdata
> collection name given as input to the digi2raw) ?
> 

My tests where what comes out of the box in runTheMatrix.py -l 140.51
If settings are not right, they should be updated in this PR.

**Comment by slava77** [comment]
@rkunnawa @abaty 
do you understand the reasons for plots failing on 2010 HI data?
If it's connected to defaults set in ::fillDescriptions, then we should have these updated to preserve the behavior. 
Is it as simple as that or something more complex going on?

**Comment by abaty** [comment]
@slava77 

The :fillDescriptions default parameters are indeed at fault for the 2010 HI data failing.  However, these parameters correct; the CopyBufferHeader parameter needs to be false in order to maintain the proper behavior for repacking MonteCarlo.  In order to recover the correct 2010 behavior, the default parameter then needs to be overwritten in the HI-specific repacking configuration in Configuration/StandardSequences/python/DigiToRaw_Repack_cff.py .  

I am preparing this change as well as another bugfix we found while looking at the 2010 test; it should be updated soon.

**Comment by abaty** [comment]
With these commits, the behavior of workflow 140.51 should be restored: 
140_51_errorperapv
140_51_errorperlumi

Also, with the bugfixes we now see FED errors in the other workflow on 2015 data I am using, which @amagnan has confirmed are in the original data stream:

hlt_test

@icali @slava77

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
140.51 now has new errors:
wf140 51_strip_apve_addbad
wf140 51_strip_apve_addr


Also, I tried to repack&rerun HLT with and without the following modifications 
```
process.hltSiStripDigiToZSRaw.CopyBufferHeader = cms.bool(True)
process.hltSiStripDigiToZSRaw.RawDataTag = cms.InputTag('rawDataCollector')
process.hltSiStripRawDigiToVirginRaw.CopyBufferHeader = cms.bool(True)
process.hltSiStripRawDigiToVirginRaw.RawDataTag = cms.InputTag('rawDataCollector')
```
and I observe some differences in tracks after that in the RECO outputs.
I used 263286  file from LS200 (HITrackerVirginRaw 56393D57-B39B-E511-A3AA-02163E011EE8.root), reran HLT without prescales and then used outputHIPhysicsSemiPeripheral as an input to RECO.
The changes are small but are still there

all_sign647-modstripvssign647_hitrackervirginrawhiphysicssemiperipheral263286c_recotracks_higeneraltracks__reco_obj_eta
all_sign647-modstripvssign647_hitrackervirginrawhiphysicssemiperipheral263286c_recotracks_higeneraltracks__reco_obj_phi
{oddly, they seem increase with abs(phi)}

If you checked something similar in your setup (both current and 2010 data), please clarify if you observe changes.

**Comment by abaty** [comment]
@slava77 
I did see very small changes in the number of tracks O(0.01%) when testing on a different HLT output stream in the 2015 data, although the new PR had slightly less tracks in that case.  I believe some small changes in the tracking are expected, as sensors with FE errors detected get automatically masked in the reconstruction sequence.

I also reproduce this 2010 APVe plot you show; I am trying to look at the input headers to see if this is actually what is in the input rawDataCollector.

**Comment by amagnan** [comment]
Concerning higher track efficiencies in the 2015 data, this is expected: now the tracking knows about modules which have errors, and ignore them when making tracks. Why it is in a specific region in eta/phi should be correlated with number of layers available for tracking (less layers -> more impact) and where are the modules which have errors all the time...
Concerning the new FEMajorityAddress errors in 2010 workflow, I'll investigate with Austin, but could I have a clarification please: you are testing the old workflow (for which DQM was running on the rawDataCollector) vs the new workflow (for which DQM is running on the rawDataRepacker with copyBufferHeader set to true), is that correct ? Thanks.

**Comment by slava77** [comment]
``` Concerning the new FEMajorityAddress errors in 2010 workflow, ```
you can see details of what I did in py files after ``` runTheMatrix.py -l 140.51 ```
This is a standard "matrix" workflow.

**Comment by slava77** [comment]
Regarding the effects on 2015 data.
I'm not particularly keen to make changes in RECO/HLT performance in the middle of a run.
- Can this effect be turned off on the raw2digi (unpacker) step when reco objects are created?
- Are these changes recoverable from the current "regular" RAW (the combination of virginRawDataRepacker and rawDataRepacker) before this PR takes effect?

This PR has clearly grown beyond just the "FED DQM error plot fix".
It's time to at least update the subject and the description of the PR.

**Comment by abaty** [comment]
@slava77 

My understanding is that this SiStrip Error information is essentially being lost forever in the currently implementation of the repacker in the HLT.  I don't think the DQM plots are recoverable without this PR.  I think @icali can comment further

**Comment by amagnan** [comment]
Indeed the FED errors are lost for ever. The other way around though, it is possible to disable the check for channels in errors during the track reconstruction, in case it would be decided that it is not wanted (which I very much doubt as the request to work on this came also from Andrea Venturi...)

**Comment by slava77** [comment]
Is there anything still to do with this PR regarding the effects on 2010 data?

**Comment by amagnan** [comment]
if I look at the errors in details, I think they are the same, just classified differently. In the new workflow, FED 59 appears with a badmajorityaddressError, whereas in the old one it was FEMissing errors. But the total number of channel in error is identical and all the other errors are identical. I have asked @abaty to rerun enabling the detailed histograms so we can confirm absolutely that they are the same channels indeed.

**Comment by slava77** [comment]
Here are a few more plots from "CopyBufferHeader =True" reHLT-reRECO in 2015 data, mainly confirming the needed features are there.
wf253286rawrep_strip_apv_errs
wf253286rawrep_strip_apve_addr
wf253286rawrep_strip_chanwerrs
wf253286rawrep_strip_err_fracs
wf253286rawrep_strip_feds_errs

Based on the recent discussion in this thread and the plots I see in 2015, I'm ready to signoff once there is a confirmation from Austin as mentioned in https://github.com/cms-sw/cmssw/pull/12674#issuecomment-163265819

**Comment by abaty** [comment]
@slava77 @icali 

After investigation with @amagnan it does appear that these are same error with different classification.  We have added a few lines of code in order to make sure that errors of the 'FEMissing' type get caught, even if they do not throw a 'badmajorityaddressError' error.  

In my testing of the 2010 and 2015 workflows, this small addition did not seem to affect the current status of the PR outputs, because of the overlap between the two error types.  If you are comfortable with signing after checking this new commit, then I think we can proceed.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by amagnan** [comment]
Just want to add that the exact error classification doesn't matter much in practice, we have checked the raw fed buffer and indeed both types could apply. I still want to understand why exactly they get reclassified but this should not prevent the PR to go ahead: it is much better to have this now than delay it any further.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/12674	8616a23
- code changes are in the packer of sistrip data in the heavy ion data taking. 
- changes introduced up to 73f0c48 are dscussed above; the last commit didn't trigger any additional changes in the monitored quantities (including matrix with 2010 HI data and reHLT-reRECO workflow of 2015 HI data)

@rkunnawa @abaty 
Please, make a pull request for 80X and also just in case for 76X.

@rkunnawa please use the "Edit" button at the top of the github page for this PR and update the subject to reflect changes made in the packer (for self-documentation)

**Comment by rkunnawa** [comment]
@slava77 done! hope we can get this now! we will prepare one for 80X and 76X as requested, with all the changes.

**Comment by Martin-Grunewald** [comment]
Is this also needed for MC production validation purposes? If so we'd need another PR from HLT as #12735 does not contain the new parameters, so would use the default setting from fillDescriptions for them, while I assume you'd need another setting for the added parameters.
Please advise!
BTW, for data I'd need to parse this and the menu needs to be migrated and then changing the added parameters. Please advise as well!

**Comment by amagnan** [comment]
Hi, no need to change anything for MC I think, default settings are fine. MC doesn't know anything about fed errors or buffer header as far as I understand....

**Comment by icali** [comment]
@Martin-Grunewald not an expert but I believe that we need to ask the parsing of the menu. This is used in only one module. I also agree with Anne Marie, nothing should be affected in MC.

**Comment by Martin-Grunewald** [comment]
Hi,

OK, I have patched it into ConfDB - code template ```CMSSW_7_5_7_patch1_HLT1```.
Default migration to this new template leads to these changes in the HIon menu:
```
-------------------------------------------------------------------------------
Modules (2):
  -> hltSiStripRawDigiToVirginRaw [SiStripDigiToRawModule] CHANGED
       bool CopyBufferHeader = false [ADDED]
       InputTag RawDataTag = rawDataCollector [ADDED]
  -> hltSiStripDigiToZSRaw [SiStripDigiToRawModule] CHANGED
       bool CopyBufferHeader = false [ADDED]
       InputTag RawDataTag = rawDataCollector [ADDED]
```
Thus, migrate the online menu you have in ConfDBv2, modified since provided by STORM,
and then adjust the new parameters in the two instances above to your liking!

**Comment by abaty** [comment]
@slava77 

We managed to run this change online today and it worked well.  I have made a PR for 8_0_X (#12762).

I have not made the PR for 7_6_X, as it has some changes that depend on having PR (#12161) merged before, and that PR seems to be held up quite a bit.  Should I just wait for that PR to merge first?

**Pull request merged**